torch/csrc/utils/python_numbers.h

```diff
@@ -41,6 +41,12 @@ inline PyObject* THPUtils_packDoubleAsInt(double value) {
 }
 
 inline bool THPUtils_checkLong(PyObject* obj) {
+#ifdef USE_NUMPY
+  if (torch::utils::is_numpy_int(obj)) {
```

Comment:
I'm a little concerned about changing the behavior of THPUtils_checkLong because it has a lot of call sites. I'm worried that it enables new behavior in a couple of places and we should figure out if we want that new behavior and add tests for them if so.

Reply:
That's a good point. For a couple of quick examples, I can see that this branch introduces the ability to: `torch.Storage(np.array([1])[0])`, where on master I get a TypeError expecting an int.  I also found I'm able to `torch.ones(5)[0]=np.array([1])[0]` on this branch but not on master.

The usage seems valid, and doesn't break any existing behavior. It allows treating `np.array([an_int])[0]` the same as `an_int` in many cases where `an_int` was considered valid int input.

I can add these couple of examples. There could be a lot of individual cases though, and I'm not sure it's worth the investment in auditing and testing every possible case.

Reply:
> I can add these couple of examples. There could be a lot of individual cases though, and I'm not sure it's worth the investment in auditing and testing every possible case.

How many callsites are there? If any of those cases crashes then we have ourselves multiple high-pri bugs in the future.

Reply:
We can also now accept np.int32 as a Long type, which is probably fine

Reply:
```
~/repos/pytorch]$ grep -Irn THPUtils_checkLong . | wc -l
63
```
Many of these are in C++ functions that back python functions which start with an `_`, e.g., THPModule_setQEngine / `_set_qengine`. We'd have to fix it if someone passed a numpy scalar and it somehow crashed, but that seems unlikely, and also not like a particularly egregious/risky potential  bug.

Reply:
We don't have to *test* every possible case, but we should at least understand the consequences of our change.

Reply:
I don't think anyone would be that surprised if stuff like this crashed for a numpy int, already crashes for `1`:

```
In [2]: torch._C._set_qengine(1)
libc++abi.dylib: terminating with uncaught exception of type c10::Error: quantized engine FBGEMM is not supported (setQEngine at ../aten/src/ATen/Context.cpp:113)
```

Reply:
It seems like anywhere int works and uses this function, numpy int should also now work. testing the first couple of cases I could find this seems to be the case (and added unit tests for a couple of the more likely-to-be-used cases).

Reply:
I looked into this a little more. Almost every use case of THPUtils_checkLong is followed by a `THPUtils_unpackLong`. As long as `THPUtils_unpackLong` supports all of the new types we are allowing in (np.(u)int64, np.(u)int32, np.(u)int16, np.(u)int8) then the callsites should be fine, but I would like to see testing for these types.

The other question then is the UI question. Should we be accepting all of these types as "Long"?

Reply:
Yes that was what I saw as well. I don't see any reason to reject a smaller or unsigned scalar when we're expecting a python int. I modified the test to check for a few np dtypes (though I didn't include uint32/64 since we don't have an unsigned 32/64 bit integral type in torch).

Reply:
So what actually happens if we try to use an unsigned numpy type with pytorch?

Reply:
it's treated as a long just like everything else

Reply:
Does np.uint64 overflow if we pass it in for a parameter like `torch.sum(x, dim=???)`? Does it error out gracefully?

Reply:
do you know of an example I could try? I don't think we can create a tensor with uint64.max > dims > int64.max

It should behave just like any int since `THPUtils_unpackLong` will do the same overflow checking it does anyway (python ints can be larger than a long)

Reply:
You don't need to create a tensor with that many dimensions to test if `torch.mean(torch.randn(1, 1), np.uint64(-1))` errors out gracefully or not. 

I don't know what PyLong_AsLongLongAndOverflow does and/or if it handles np.uint64. PyLong should only work on PyLong and subclasses of it (the np integer types aren't actually subclasses of PyLong...) so we should test the behavior.

Reply:
I think the np integer types are subclasses of PyLong (by setting the Py_TPFLAGS_LONG_SUBCLASS type flag in their custom type).

Checking, `torch.mean(torch.randn(1, 1), np.uint64(-1))` gives `RuntimeError: Overflow when unpacking long`
on this branch as I'd expect. This does differ from master which gives `TypeError: mean() received an invalid combination of arguments - got (Tensor, numpy.uint64), but expected one of:...`

Reply:
Sounds good, can you add that as a test case?

Reply:
added